### PR TITLE
snap-bootstrap: add tpm support (draft)

### PR DESCRIPTION
**Pull request by cmatsuoka**
Detect, provision and use the TPM device (if available) to seal the encrypted partition key, and add the `open-encrypted` command to snap-bootstrap to unseal the key and open the encrypted device.

To open the encrypted device, invoke snap-bootstrap as:
```
/usr/lib/snapd/snap-bootstrap open-encrypted -o insecure-tpm-connection ubuntu-data /dev/vda4 /run/mnt/ubuntu-boot/keyfile
```
The TPM commits are on top of PR #7999 and the recovery key branch. Tests still missing and will be added in future commits.

### CODE REVIEW

File: cmd/snap-bootstrap/partition/encrypt.go [hunk 1]
```diff
@@ -52,7 +58,43 @@ func (key EncryptionKey) Store(filename string) error {
...
+	// 5 digits being converted to a 2-byte number to make a 16-byte key.
+	r := bytes.NewReader(key[:])
+	for i := 0; i < 8; i++ {
+		var val uint16
+		if err := binary.Read(r, binary.LittleEndian, &val); err != nil {
+			return err
+		}
+		groups[i] = fmt.Sprintf("%05d", val)
+	}
+
```
Comment (chrisccoulson) [inline cmd/snap-bootstrap/partition/encrypt.go]: It's fine to just store the 16-byte key in binary format here. If we introduce any tooling later on for backing up the recovery key, the conversion to the numeric version can probably happen there if it needs to.

Comment (cmatsuoka) [reply cmd/snap-bootstrap/partition/encrypt.go]: It would also be useful to have something to regenerate the key in case it's compromised.

Comment (cmatsuoka) [reply cmd/snap-bootstrap/partition/encrypt.go]: Fixed this in PR #8103. I'll rebase on top of the new recovery key branch once it gets more reviews.

**Comment by xnox** [comment]
This branch fails to cleanbuild with:

```
src/github.com/snapcore/snapd/vendor/github.com/chrisccoulson/go-tpm2/device_linux.go:12:2: cannot find package "golang.org/x/sys/unix" in any of:
	/root/build_snapd/parts/snapd/build/_build/src/github.com/snapcore/snapd/vendor/golang.org/x/sys/unix (vendor tree)
	/usr/lib/go-1.10/src/golang.org/x/sys/unix (from $GOROOT)
	/root/build_snapd/parts/snapd/build/_build/src/golang.org/x/sys/unix (from $GOPATH)
```

are we missing vendoraziation of some packages? or do i need to rerun govendor something?

**Comment by cmatsuoka** [comment]
This PR was replaced by PR #8409

**Pull request closed without merging**